hpx/parallel/algorithms/sort.hpp

```diff
@@ -0,0 +1,271 @@
+//  Copyright (c) 2015 John Biddiscombe
+//  Copyright (c) 2015 Hartmut Kaiser
+//  Copyright (c) 2015 Francisco José Tapia
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#if !defined(HPX_PARALLEL_ALGORITHM_SORT_OCT_2015)
+#define HPX_PARALLEL_ALGORITHM_SORT_OCT_2015
+
+#include <hpx/config.hpp>
+#include <hpx/traits/concepts.hpp>
+
+#include <hpx/parallel/config/inline_namespace.hpp>
+#include <hpx/parallel/execution_policy.hpp>
+#include <hpx/parallel/algorithms/detail/dispatch.hpp>
+#include <hpx/parallel/algorithms/detail/is_negative.hpp>
+#include <hpx/parallel/util/detail/algorithm_result.hpp>
+#include <hpx/parallel/util/projection_identity.hpp>
+#include <hpx/parallel/traits/projected.hpp>
+#include <hpx/lcos/local/dataflow.hpp>
+
+#include <algorithm>
+#include <iterator>
+#include <type_traits>
+
+namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
+{
+
+    ///////////////////////////////////////////////////////////////////////////
+    // sort
+    namespace detail
+    {
+        /// \cond NOINTERNAL
+
+        ///---------------------------------------------------------------------------
+        /// @struct iter_value
+        /// @brief This is to obtain the type of data pointed to by an iterator
+        /// @tparam RandomIt type of iterator
+        /// @remarks
+        //----------------------------------------------------------------------------
+        template <typename RandomIt>
+        struct iter_value
+        {
+            typedef typename
+                std::remove_reference<decltype(*(std::declval<RandomIt>()))>::type type;
+        };
```

Comment:
Looks like, this could be replaced by `std::iterator_traits<RandomIt>::value_type`.